SEP-2640: Skills Extension (#2640)
modelcontextprotocol/modelcontextprotocol: seps/2640-skills-extension.md

Further constraints:

- A `SKILL.md` MUST NOT appear in any descendant directory of a skill. The skill directory is the boundary; skills do not nest inside other skills.
- The final `<skill-path>` segment, being the skill `name`, MUST satisfy the Agent Skills specification's naming rules. Prefix segments SHOULD be valid URI path segments per [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986); no further constraints are imposed on them.

Per RFC 3986, the first segment of `<skill-path>` occupies the authority component. This carries no special semantics under this convention and clients MUST NOT attempt DNS or network resolution of it.

> A server MAY instead serve skills under another scheme native to its domain (e.g., `github://owner/repo/skills/refunds/SKILL.md`), provided each skill is listed in the [`skill://index.json`](#enumeration-via-skillindexjson) resource. The index is the authoritative record of which resources are skills; outside the index, hosts recognize skills by the `skill://` scheme prefix. The structural constraints above — `<skill-path>` ending in the skill name, `SKILL.md` explicit in the URI, no nesting — apply regardless of scheme.

Reviewer: So the skill path for the `github:// example` is `owner/repo/skills/refunds`, right?

Author: Yes

## Implementation Guidelines

The following are recommendations for interoperable implementations. They are not part of the normative specification.

### Hosts: Model-Driven Resource Loading

Hosts SHOULD expose a tool to the model that reads MCP resources by server and URI, enabling the model to load skill content on demand:

```json
{
  "name": "read_resource",
  "description": "Read an MCP resource from a connected server.",
  "inputSchema": {
    "type": "object",
    "properties": {
      "server": {
        "type": "string",
        "description": "Name of the connected MCP server"
      },
      "uri": {
        "type": "string",
        "description": "The resource URI, e.g. skill://git-workflow/SKILL.md"
      }
    },
    "required": ["server", "uri"]
  }
}

> ```

Reviewer: I know this is just an example - but I don't think the model typically knows about MCP servers in most client implementations (or do they in your investigations into existing read_resource impls?).

For tools, the host keeps track of which tools exposed to the model come from MCP servers; it doesn't disclose to the model which MCP servers are available or how tools map to individual MCP servers.

Similarly, I would expect hosts to potentially have something like a single read_skill progressive disclosure tool that can read skills locally on disk and from MCP servers.  Especially wrt the section below "Unified Treatement of Filesystem and MCP Skills". It's not clear to me how this signature would support unified treatment, since local file-based skills wouldn't have a `server` input value.

Author: @olaservo did an exploration here: https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2527#issuecomment-4281339798

I did some digging in code and:
* claude code, codex, and fast-agent all have a single `read_resource` tool with a server param.
* gemini-cli and goose just take uri alone and seem to probe servers one-by-one until one hits... which seems pretty janky and IMO should be fixed regardless of skills.

but yes, it's non-normative and implementations are free to do what they like here.

> Similarly, I would expect hosts to potentially have something like a single read_skill progressive disclosure tool that can read skills locally on disk and from MCP servers. Especially wrt the section below "Unified Treatement of Filesystem and MCP Skills". It's not clear to me how this signature would support unified treatment, since local file-based skills wouldn't have a server input value.

What I expect (but again, non-normative):

* Client enumerate skills for all servers by reading `skill://index.json`
* They create a map of skill-name to skill metadata. The skill metadata says whether it is a local file or MCP-backed skill + which server it came from (or anything else).
* They have a single `read_skill` tool that keys on skill-name and when invoked uses the metadata to decide whether to do a local file read or MCP `resources/read` (note: the model doesn't need to use `read_resource` for this -- just `read_skill` and it is backed by a `resource/read` for MCP skills).
* `read_resource` is mostly needed for reading related files as part of the skill, e.g. if the skill references `skill://code-review/checklist.json` then the model will need to read that using a `read_resource` tool and make sure it is disambiguated.

Reviewer: That matches how I'm thinking about too, I recommend adding it to the SEP

Author: Done!

- **Skills are data, not directives.** Hosts MUST NOT treat skill resources as higher-authority than other context. Explicit user policy governs whether a skill is loaded at all.
- **Provenance and inspection.** Hosts SHOULD indicate which server a skill originates from when presenting it, SHOULD let users inspect a skill's content before it is loaded into model context, and MAY gate loading behind per-skill or per-server user approval.
- **Archive safety.** Archive-distributed skills are an unpacking attack surface. Hosts MUST validate archives per the [Agent Skills archive safety requirements](https://agentskills.io/well-known-uri#archive-safety): reject path traversal and absolute paths, reject links resolving outside the skill directory, and bound total unpacked size to prevent decompression bombs.
- **Not a third-party marketplace.** This extension is for servers to ship skills that describe their own tools, not for distributing arbitrary third-party content through a connected server.

## Reference Implementation

Per [SEP-2133], an Extensions Track SEP requires at least one reference implementation in an official SDK prior to review.

**SDK implementation:**

> - TypeScript SDK convenience wrappers (`@server.skill()`, `client.listSkills()`, `client.readSkillUri()`): [experimental-ext-skills#71](https://github.com/modelcontextprotocol/experimental-ext-skills/pull/71)

Reviewer 2: I'm second guessing if we should create this, since there isn't new protocol surface being added and so publishing and maintaining convenience wrappers seems like overkill.  I think some combination of the SEP itself and a brief implementation guide pointing to examples should be enough?

Author: sorry, when you say "create this", do you mean:

- SDKs shouldn't provide these convenience wrappers
- they should, but *we* shouldn't publish them?

From my perspective:
- I do think the SDKs should have these wrappers since it just makes it easier to maintain skills, especially not having to manually create an index.json
- As for us, we should have a reference implementation in this PR (it's a requirement). Maybe the SDK maintainers would prefer to build it themselves, but we should have something to point to. We do need to land something ahead of finalization though.

(I'm probably misunderstanding you though)

Reviewer 2: I just meant do we need the helpers at all, given that they aren't doing anything very complicated, but yes it makes sense to streamline the boilerplate stuff.

## Motivation

Native skills support in host applications demonstrates strong demand for rich, progressively disclosed workflow instructions. MCP does not currently offer a conventional way to ship this content alongside the tools it describes, which leads to:

- **Fragmented distribution.** A server and the skill that teaches an agent to use it are versioned, discovered, and installed separately. Users installing a server from a registry have no signal that a companion skill exists. ([problem statement](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/problem-statement.md))
- **Instruction size limits.** Server instructions load once at initialization and are practically bounded in size. Complex workflows — such as the 875-line [mcpGraph skill](https://github.com/TeamSparkAI/mcpGraph/blob/main/skills/mcpgraphtoolkit/SKILL.md) — do not fit this model. ([experimental findings](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/experimental-findings.md#mcpgraph-skills-in-mcp-server-repo))

> - **Inconsistent ad-hoc solutions.** Absent a convention, four independent implementations have each invented their own `skill://` URI structure, with diverging semantics for authority, path, and sub-resource addressing.

Reviewer 2: ```suggestion
- **Inconsistent ad-hoc solutions.** Absent a convention, several independent implementations have each invented their own `skill://` URI structure, with diverging semantics for authority, path, and sub-resource addressing.
```

There are more than 4 examples in the wild so I think we can make this indicate that.

- `SKILL.md` MUST begin with YAML frontmatter containing at minimum the `name` and `description` fields as defined by the Agent Skills specification.
- A skill MAY contain additional files and subdirectories (references, scripts, examples, assets).

This extension does not redefine, constrain, or extend the skill format. Future revisions of the Agent Skills specification apply automatically.

### Resource Mapping

Each file within a skill directory is exposed as an MCP resource. Servers SHOULD use the `skill://` URI scheme, under which the resource URI has the form:

```
skill://<skill-path>/<file-path>
```

A server MAY instead serve skills under another scheme native to its domain (e.g., `github://owner/repo/skills/refunds/SKILL.md`), provided each skill is listed in the [`skill://index.json`](#enumeration-via-skillindexjson) resource. The index is the authoritative record of which resources are skills; outside the index, hosts recognize skills by the `skill://` scheme prefix.

The structural constraints below — `<skill-path>` ending in the skill name, `SKILL.md` explicit in the URI, no nesting — apply regardless of scheme.

> where:

Reviewer 2: Is this a fragment left from editing?

- `<file-path>` is the file's path relative to the skill directory root, using `/` as the separator.

The resource for the skill's required `SKILL.md` is therefore always addressable as `skill://<skill-path>/SKILL.md`, and the skill's root directory is the URI obtained by stripping the trailing `SKILL.md`.

The final segment of `<skill-path>` MUST equal the skill's `name` as declared in its `SKILL.md` frontmatter. This mirrors the Agent Skills specification's requirement that `name` [match the parent directory name](https://agentskills.io/specification#name-field). Preceding segments, if any, are a server-chosen organizational prefix — servers MAY organize skills hierarchically by domain, team, version, or any other axis. In `skill://acme/billing/refunds/SKILL.md`, the prefix is `acme/billing` and the skill's `name` is `refunds`; in `skill://git-workflow/SKILL.md` there is no prefix and the `name` is `git-workflow`. This means the skill name is always recoverable from the URI alone, without reading frontmatter.

Further constraints:

> - A `SKILL.md` MUST NOT appear in any descendant directory of a skill. The skill directory is the boundary; skills do not nest inside other skills.

Reviewer 2: Sanity checking this: @jonathanhefner is this restriction consistent with how skills in general should or shouldn't reference other skill files in a nested way, independent of this SEP?

Reviewer 2: Checked the agent skills spec, and while it doesn't explicitly forbid nested SKILL.md its model (one SKILL.md per directory, parent-dir-name = name, no skill-to-skill reference notion, 'avoid deeply nested reference chains') is consistent with our restriction. Going to resolve this comment.

#### Examples

| Skill path             | File                  | Resource URI                                      |
| ---------------------- | --------------------- | ------------------------------------------------- |
| `git-workflow`         | `SKILL.md`            | `skill://git-workflow/SKILL.md`                   |
| `pdf-processing`       | `references/FORMS.md` | `skill://pdf-processing/references/FORMS.md`      |
| `pdf-processing`       | `scripts/extract.py`  | `skill://pdf-processing/scripts/extract.py`       |
| `acme/billing/refunds` | `SKILL.md`            | `skill://acme/billing/refunds/SKILL.md`           |
| `acme/billing/refunds` | `templates/email.md`  | `skill://acme/billing/refunds/templates/email.md` |

#### Resource Metadata

For each `skill://<skill-path>/SKILL.md` resource:

- `mimeType` SHOULD be `text/markdown`.
- `name` SHOULD be set from the `name` field of the `SKILL.md` YAML frontmatter. By the path constraint above, this will equal the final segment of `<skill-path>`.
- `description` SHOULD be set from the `description` field of the `SKILL.md` YAML frontmatter.

> Servers MAY expose additional frontmatter fields via the resource's `_meta` object. Other files in the skill use the `mimeType` appropriate to their content.

Reviewer 2: Should we add:

`When _meta keys are needed for skill resources, implementations SHOULD use the io.modelcontextprotocol.skills/ reverse-domain prefix.`

And potentially link to (or include) other [documentation](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/skill-meta-keys.md) around `_meta` usage for skill resources?

Hosts SHOULD load the frontmatter (`name`, `description`) of available and enabled skills into the model's context so the model can judge relevance and construct a `read_resource` call when a skill applies. Hosts SHOULD surface available skills in their UI for user inspection and per-skill enable/disable, analogous to how tools are typically exposed.

A typical flow: the host reads `skill://index.json` from each connected server and surfaces the `name` and `description` of each entry in the model's context. The model calls `read_resource` with a concrete URI — one returned by enumeration, one handed to it by the user (who may have resolved it from a template entry in the host UI), or one obtained out-of-band — when a skill is relevant to the task.

Because enumeration is optional, a `read_resource` call for a `skill://` URI that the host has never seen listed is normal and expected. The host forwards it to the named server; the server either serves the resource or returns a not-found error.

> ### Hosts: Unified Treatment of Filesystem and MCP Skills

Reviewer 2: I think this one should come first?

```python
@server.skill("git-workflow")                 # → skill://git-workflow/SKILL.md
def git_workflow():
    return Path("./skills/git-workflow")

@server.skill("acme/billing/refunds")         # → skill://acme/billing/refunds/SKILL.md
def refunds():
    return Path("./skills/refunds")
```

The SDK handles: reading `SKILL.md` frontmatter to populate resource metadata, registering a `skill://<skill-path>/{+path}` resource template, serving file content on `resources/read`, and (where the server's skill set is bounded) generating the `skill://index.json` resource.

**Client-side** — enumerate and fetch skills:

```python
skills = await client.list_skills()               # reads skill://index.json, may be empty or absent
content = await client.read_skill_uri(
    "skill://acme/billing/refunds/SKILL.md")      # wraps resources/read, works regardless of enumeration
```

These wrappers are thin — each is a single underlying protocol call with a fixed URI pattern — but they give server authors an ergonomic way to declare skills and give client authors a discoverable entry point.

## Rationale

### Why Resources Instead of a New Primitive?

> The Working Group's [decision log](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/decisions.md#2026-02-26-prioritize-skills-as-resources-with-client-helper-tools) records this as settled. Skills are files; Resources exist to expose files. Reusing Resources inherits URI addressability, `resources/read`, `resources/subscribe`, templates, and the existing client tooling for free. A new primitive would duplicate most of this and add ecosystem complexity — a concern raised in [community discussion](https://github.com/modelcontextprotocol/experimental-ext-skills/issues/14).

Reviewer 2: ```suggestion
The Working Group's [decision log](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/decisions.md#2026-02-26-prioritize-skills-as-resources-with-client-helper-tools) records this as settled. Skills are files; Resources exist to expose files. Reusing Resources inherits URI addressability, `resources/read`, `resources/subscribe`, templates, and the existing client tooling for free. A new primitive would duplicate most of this and add ecosystem complexity.
```

I think we can leave out linking to this issue.  It seems like a safe assumption that adding new primitives and mechanisms to the protocol will add complexity.  Using resources to describe files is also aligned with [Composability over specificity](https://modelcontextprotocol.io/community/design-principles#composability-over-specificity) and other MCP [design principles](https://modelcontextprotocol.io/community/design-principles).

> [SEP-2076] proposes the new-primitive alternative. That approach offers cleaner capability negotiation and dedicated list-changed notifications, but at the cost of flattening skills to name-addressed blobs — losing the directory model that the Agent Skills specification defines and that supporting files depend on.

Reviewer 2: ```suggestion
[SEP-2076] originally proposed the new-primitive alternative. That approach offers cleaner capability negotiation and dedicated list-changed notifications, but at the cost of flattening skills to name-addressed blobs — losing the directory model that the Agent Skills specification defines and that supporting files depend on.
```

### Why `skill://<path>/<file>` With an Explicit `SKILL.md`?

> Four independent implementations converged on `skill://` as the scheme without coordination — a strong signal. They diverged on structure. This SEP adopts the explicit-file form because:

Reviewer 2: ```suggestion
Several independent implementations converged on `skill://` as the scheme without coordination — a strong signal. They diverged on structure. This SEP adopts the explicit-file form because:
```

### Why Allow a Path Prefix But Constrain the Final Segment?

Earlier drafts required `<skill-path>` to be a single segment equal to the frontmatter `name`. That breaks down when a server needs hierarchy: an organization serving both `acme/billing/refunds` and `acme/support/refunds` cannot satisfy "single segment" without renaming one skill to dodge the collision. Allowing a prefix (`acme/billing/`, `acme/support/`) solves this — both skills can be named `refunds` and the prefix disambiguates.

A subsequent draft went further and fully decoupled the path from the name. That was too loose: a URI like `skill://a/b/c/SKILL.md` tells you nothing about what the skill is called until you fetch and parse frontmatter. Clients listing skills, hosts displaying them in a picker, and models reasoning over URIs all want the name visible without a round trip.

Constraining the final segment to match the frontmatter `name` gets both properties. The prefix carries the server's organizational structure; the final segment carries the skill's identity; and the two together form a locator from which the name can be read directly.

> ### Why Is Enumeration Optional?

Reviewer 2: ```suggestion
### Why Is Enumeration by Index Optional?
```

Reviewer 2: (Or, was this wording intentional to reflect the fact that resource templates can be used?)

Author: Intentional — the section covers why enumeration in general is optional, which includes both the `index.json` and resource-template paths. Adding "by Index" would narrow it to one mechanism. Keeping as-is.

Reinstating archives as a server-side packaging choice — not a client-visible mode split — preserves the virtual-filesystem model: the post-unpack view is identical to individual-file distribution. The property given up is per-file `resources/subscribe` granularity, which is acceptable because subscription is not part of the skill reading model defined here.

## Backward Compatibility

This extension introduces no new protocol methods, message types, or schema changes. A server that does not implement this extension simply exposes no `skill://` resources; existing clients are unaffected. A client that does not implement this extension sees `skill://` resources as ordinary resources, which they are.

> Existing implementations using other `skill://` URI structures (NimbleBrain's `skill://server/skill`, skilljack's implicit-`SKILL.md` `skill://name`) will need to adjust their URI paths to conform. These are small, mechanical changes.

Reviewer 2: One significant example we could call out here is the FastMCP [SkillsProvider](https://gofastmcp.com/servers/providers/skills).

Examples of differences in how FastMCP skill resources are implemented:

- URI/identity: name-segment not validated against frontmatter, no hierarchical paths.
- Discovery: uses a per-skill `skill://{name}/_manifest` JSON listing resource
- Metadata: Resource `name` field set to `{skill}/SKILL.md` rather than frontmatter `name`.

I think these plus other examples help make the argument for standardization more than they argue against proposing the conventions in the SEP, but I think this is one 'breaking' example we will want to proactively follow up on sooner than later given the popularity of FastMCP.

Also don't think we should just refer to the skilljack server by name here since it's not going to be broadly recognizable.  I think we could point to an open-ended survey of real implementations in the WG repo by linking to [this](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/related-work.md)?

- **Skills are data, not directives.** Hosts MUST NOT treat skill resources as higher-authority than other context. Explicit user policy governs whether a skill is loaded at all.
- **Provenance and inspection.** Hosts SHOULD indicate which server a skill originates from when presenting it, SHOULD let users inspect a skill's content before it is loaded into model context, and MAY gate loading behind per-skill or per-server user approval.
- **Archive safety.** Archive-distributed skills are an unpacking attack surface. Hosts MUST validate archives per the [Agent Skills archive safety requirements](https://agentskills.io/well-known-uri#archive-safety): reject path traversal and absolute paths, reject links resolving outside the skill directory, and bound total unpacked size to prevent decompression bombs.
- **Not a third-party marketplace.** This extension is for servers to ship skills that describe their own tools, not for distributing arbitrary third-party content through a connected server.

> The instructor-only scope of this extension ([decisions.md, 2026-02-14](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/decisions.md#2026-02-14-skills-served-over-mcp-use-the-instructor-format)) deliberately excludes the helper model. A filesystem skill might reasonably carry scripts the user has audited; an MCP skill arrives from a remote party and MUST be handled as text that influences model behavior, not as code that executes on the host.

Reviewer 2: Not sure we need this extra paragraph since it seems redundant with the other points above?

Author: Agreed on reflection — the rules in the bullets cover it and the decision log is already linked from References. Removed in d907276b.

## Reference Implementation

Per [SEP-2133], an Extensions Track SEP requires at least one reference implementation in an official SDK prior to review.

**SDK implementation:**

- TypeScript SDK convenience wrappers (`@server.skill()`, `client.listSkills()`, `client.readSkillUri()`): [experimental-ext-skills#71](https://github.com/modelcontextprotocol/experimental-ext-skills/pull/71)

> **Host implementations** (reading `skill://` resources, surfacing skills alongside filesystem skills):

Reviewer 2: ```suggestion
**Prototype host implementations** (reading `skill://` resources, surfacing skills alongside filesystem skills):
```

(Just to make it obvious that these are for early testing and not finalized PRs.)

- gemini-cli: [olaservo/gemini-cli#1](https://github.com/olaservo/gemini-cli/pull/1)
- fast-agent: [olaservo/fast-agent#1](https://github.com/olaservo/fast-agent/pull/1)
- goose: [olaservo/goose#1](https://github.com/olaservo/goose/pull/1)
- codex: [olaservo/codex#1](https://github.com/olaservo/codex/pull/1)
- Claude Code: prototyped internally at Anthropic; not yet public

> **Server implementation:**

Reviewer 2: ```suggestion
**Prototype server implementation:**
```

(In general the intention is to expand this section for both hosts and servers in the coming weeks.)

- GitHub MCP Server: [github/github-mcp-server#2360](https://github.com/github/github-mcp-server/pull/2360)

## References

- [Agent Skills specification](https://agentskills.io/specification)
- [Agent Skills well-known URI discovery](https://agentskills.io/well-known-uri)
- [SEP-2133]: Extensions
- [SEP-2076]: Agent Skills as first-class primitive (alternative approach)
- [Skills Over MCP Working Group charter](https://modelcontextprotocol.io/community/skills-over-mcp/charter)
- [Decision Log](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/decisions.md) — Working Group decisions and rationale

> - [Experimental Findings](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/experimental-findings.md) — results from implementations

Reviewer 2: ```suggestion
- [Experimental Findings](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/experimental-findings.md) — results from implementations (WIP)
```

Since right now it doesn't have any of the results from this SEP yet (PR is incoming).

| `git-workflow`         | `SKILL.md`            | `skill://git-workflow/SKILL.md`                   |
| `pdf-processing`       | `references/FORMS.md` | `skill://pdf-processing/references/FORMS.md`      |
| `pdf-processing`       | `scripts/extract.py`  | `skill://pdf-processing/scripts/extract.py`       |
| `acme/billing/refunds` | `SKILL.md`            | `skill://acme/billing/refunds/SKILL.md`           |
| `acme/billing/refunds` | `templates/email.md`  | `skill://acme/billing/refunds/templates/email.md` |

#### Resource Metadata

For each `skill://<skill-path>/SKILL.md` resource:

- `mimeType` SHOULD be `text/markdown`.
- `name` SHOULD be set from the `name` field of the `SKILL.md` YAML frontmatter. By the path constraint above, this will equal the final segment of `<skill-path>`.
- `description` SHOULD be set from the `description` field of the `SKILL.md` YAML frontmatter.

Servers MAY expose additional frontmatter fields via the resource's `_meta` object. When `_meta` keys are used for skill resources, implementations SHOULD use the `io.modelcontextprotocol.skills/` reverse-domain prefix. Other files in the skill use the `mimeType` appropriate to their content.

> ### Discovery

Reviewer 3: This model works great for bundling skills related to an mcp server to the same mcp server.

How do we envision shipping skills via MCP when skills describe multi step workflows which involve tool invocations from multiple mcp servers?

1. To mitigate, should I create a separate MCP server just for skill discovery?
2. If so, can skills bundled with mcpServer1 reference also tools from mcpServer2?

Author: There's not a clean answer for this right now, but a few thoughts:

* Skills are just text inserted into context, so you can reference whatever servers you like. It's on the user to make sure they're all installed and available though.
* Yes, one option could be to create a separate MCP server for this. Or you don't necessarily need to use skills-over-MCP for multi-server skills.
* There have been discussions around having skills denote dependencies in the YAML frontmatter. I don't believe anything got standardized here though, but could be a path in the future.

A server is not required to make its skills enumerable. A skill's URI is directly readable via `resources/read` whether or not it appears in any index, and hosts MUST support loading a skill given only its URI (see [Hosts: End-to-End Integration](#hosts-end-to-end-integration)). This is the baseline: if a model has the URI — from server instructions, from another skill, from the user — it can read the skill.

> On top of that baseline, two discovery mechanisms are defined. A server MAY support either or both.

Reviewer 4: Isn't there an implicit third mechanism which is to simply recursively enumerate the `skills://` resource tree?  The SEP seems to be worded that SKILL.md files have to be specified in either index.json or server instructions(?) but I don't see why that's the case?

Author: There is no resource tree (it's a flat list) so I assume here you mean just listing all resources to find skills?

I think we want to avoid proposing that. Servers may have thousands (or millions) of resources and without a way to filter to skills you be doing a lot of unnecessary work on startup. It would be like coding agents crawling your entire disk to find skills instead of relying on known locations.

Reviewer 4: Sorry missed this comment at the time; I agree [scanning] is undesirable !

#### Enumeration via `skill://index.json`

> A server SHOULD expose a resource at the well-known URI `skill://index.json` whose content is a JSON index of the skills it serves. The index format follows the [Agent Skills well-known URI discovery index](https://agentskills.io/well-known-uri#index-format), with two differences: the `url` field contains a full MCP resource URI (any scheme the server serves), and the `digest` field is omitted (integrity is the transport's concern over an authenticated MCP connection). This binding also defines one additional `type` value, `"mcp-resource-template"`, for entries that describe a parameterized skill namespace.

Reviewer 4: This link is currently a 404. Suggested alternatives for reviewers:
https://github.com/agentskills/agentskills/pull/254
https://github.com/cloudflare/agent-skills-discovery-rfc

Clients SHOULD ignore unrecognized fields and SHOULD skip entries with an unrecognized `type`.

**Skill-md entries** (`type: "skill-md"`) point `url` at the full resource URI of the skill's `SKILL.md`. Supporting files are individually addressable as sibling resources under the same skill path.

**Archive entries** (`type: "archive"`) point `url` at a single resource whose content is a packed skill directory. Per the [Agent Skills archive distribution format](https://agentskills.io/well-known-uri#archive-distribution), the archive MUST be `.tar.gz` (gzip-compressed tar, `mimeType` `application/gzip`) or `.zip` (`mimeType` `application/zip`); hosts MUST support both and SHOULD determine the format from the resource's `mimeType`, falling back to the URL suffix. Archive contents represent the skill directory directly — `SKILL.md` MUST be at the archive root, not nested inside a wrapper directory, and the archive MUST NOT contain path-traversal sequences (`..`) or absolute paths.

> The `<skill-path>` is the archive `url` with its archive suffix (`.tar.gz` or `.zip`) stripped — `skill://pdf-processing.tar.gz` unpacks to `skill://pdf-processing/`, `skill://acme/billing/refunds.zip` to `skill://acme/billing/refunds/` — and the archive contents populate `skill://<skill-path>/<file-path>` exactly as if each file were served individually. A server chooses per-skill whether to distribute as individual files or as an archive; hosts observe an identical virtual namespace either way. Archive distribution delivers a multi-file skill atomically in one round trip and can carry UNIX file metadata (executable bits, symlinks) that individual-resource distribution cannot represent — robustly for `.tar.gz`, variably for `.zip`.

Reviewer 5: Does the index.json have a constraint on unique way to expose the skill?

For instance, if I have a skill folder, few agents may prefer to download the archive at startup time, while other agents would prefer to load in only the skill.md file and load in related resources at runtime.

Can we expose the skill as "skill-md" and "archive" resource types to let agent/harness know about the two styles? Any other alternatives for this

Reviewer 6: Similar to my request at https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2640#issuecomment-4480173794

Reviewer 5: Thanks for linking that. We are trying to implement this spec in Microsoft Foundry Toolbox as well to expose skills as resources. As part of that we want to explore both methods of distribution:

1. discover skills (single file + folders), download them at agent session init time
2. discover skills (single file + folders), but only download the resources into the agent sandbox/session as needed.
    - Retrieve only the skill://index.json at init time
    - If an "example skill" needs to be activated, then resource/read on `skill://example_skill/SKILL.md`
    - If the SKILL.md references other documents (doc X.md), then do a resource/read on that `skill://example_skill/X.md`

Though downloading the archive on session init might probably be the best for long-lived sessions. There are cases where if my skill is massive folder of instructions (>10MB), we can just download few files we need and not the entire archive. With size hints provided, the agent/harness can make a determination on choosing to download archive or only certain files.

The skills spec https://agentskills.io/client-implementation/adding-skills-support assumes the agent is reading from file system (so archive makes immediate sense). But there is a growing use case for having progressive disclosure over the network than even having all skills locally in file system (especially in case of sandboxes with limited memory and in cases where there are just too many skills). Perhaps this is best delegated to agent skills spec itself.

For now, we are experimenting by exposing a skill folder in index.json in 2 ways (one skill-md entry and one archive entry), so the agent/harness can use the approach of their choice. Ideally if each skill could have offer a list of possible discovery and load mechanism, it would be nicer but this likely requires alignment with agentskills.io spec as well.

#### Capability Declaration

Per [SEP-2133] extension negotiation, servers declare support for this extension in their `initialize` response:

```json
{
  "capabilities": {
    "extensions": {
      "io.modelcontextprotocol/skills": {}
    }
  }
}
```

No extension-specific settings are currently defined; an empty object indicates support.

### Reading

Skill files are read via the standard `resources/read` method. No skill-specific read semantics are defined.

Internal references within a skill (e.g., `SKILL.md` linking to `references/GUIDE.md`) are relative paths, as in the filesystem form of the Agent Skills specification. A client resolves a relative reference against the skill's root — `references/GUIDE.md` in `skill://acme/billing/refunds/SKILL.md` resolves to `skill://acme/billing/refunds/references/GUIDE.md` — exactly as a filesystem path would resolve. The skill's root is the directory containing `SKILL.md`, not the `skill://` scheme root.

## Implementation Guidelines

The following are recommendations for interoperable implementations. They are not part of the normative specification.

> ### Hosts: End-to-End Integration

Reviewer 4: My honest feedback is that this SEP would be much better without this Hosts: End-to-End Integration section. The goal "... from the model's perspective, an MCP-served skill is indistinguishable from a filesystem skill." is useful, and https://agentskills.io/client-implementation/adding-skills-support is a good reference.

The main question I have as a host implementor is the expected behaviour for `mcp-resource-template`. Is the "addressable skill space" in the description intended for the model. Hydrating and invoking this class of Skills is unaddressed.

Author: > My honest feedback is that this SEP would be much better without this

Fair. I also received feedback earlier that people wanted more detail on how to implement this in hosts, so hearing conflicting feedback here. Leaning towards leaving it unless it is doing specific damage.

> The main question I have as a host implementor is the expected behaviour for mcp-resource-template. Is the "addressable skill space" in the description intended for the model. Hydrating and invoking this class of Skills is unaddressed.

The way I've implemented it is to not provide it to the model. Instead, the skills are just presented via UX (e.g. user types `/my-mcp-server:` and is presented with completion-backed auto complete suggestions for skill templates). In that case, the index description is only presented to the user via UX during completion.

That being said, I could see a world where you provide completions to the model and present a template to the model as well so that the model could decide to explore a template autonomously based on the broad description. Probably a very niche use case though.

Reviewer 7: I had similar thoughts on using completions for skill discovery / enumeration. I wonder if something on completions deserves a section in the spec or maybe even a top-level alternative to `index.json`

- **Skills are data, not directives.** Hosts MUST NOT treat skill resources as higher-authority than other context. Explicit user policy governs whether a skill is loaded at all.
- **Provenance and inspection.** Hosts SHOULD indicate which server a skill originates from when presenting it, SHOULD let users inspect a skill's content before it is loaded into model context, and MAY gate loading behind per-skill or per-server user approval.
- **Archive safety.** Archive-distributed skills are an unpacking attack surface. Hosts MUST validate archives per the [Agent Skills archive safety requirements](https://agentskills.io/well-known-uri#archive-safety): reject path traversal and absolute paths, reject links resolving outside the skill directory, and bound total unpacked size to prevent decompression bombs.

> - **Not a third-party marketplace.** This extension is for servers to ship skills that describe their own tools, not for distributing arbitrary third-party content through a connected server.

Reviewer 2: If read literally, I think this bullet excludes legitimate patterns the SEP itself endorses (like the GitHub MCP Server using resource templates for dynamic skill discovery from repos, where repo authors write the skills and the server is just the transport).

It seems like the concern the bullet is meant to address is servers relaying resources that aren't coming directly from the server author, in a way that isn't transparent to the skill consumer?  So maybe we should add something like this instead:

> **Surface third-party authorship.** When a server relays skills it did not author (e.g., a registry surfacing skills shipped in published packages), it MUST surface that authorship to the host so the user can see whose content they're loading.

@SamMorrowDrums what do you think?  In the GitHub example, the URI includes the github user or org and the repo, which communicates the source of the file.  And the description of the resource template could indicate this detail to the client/model reading the template. If that makes sense, we could include that as a more concrete example of 'surfacing authorship.'

Reviewer 2: This bullet was removed, going to close this comment and take the related convo to the discord instead.

The `frontmatter` object MUST be identical in content to the frontmatter of the `SKILL.md` it describes. The final `<skill-path>` segment of the entry's `url` MUST equal `frontmatter.name`, per [Resource Mapping](#resource-mapping).

Within the frontmatter `metadata` object, keys prefixed with `io.modelcontextprotocol/` are reserved for metadata defined by MCP extensions. This extension currently defines no such keys. Implementations SHOULD ignore keys under this prefix that they do not recognize.

##### Skill URLs

`url` is the full resource URI of the skill's `SKILL.md`, readable via `resources/read`. Supporting files are individually addressable as sibling resources under the same skill path, per [Resource Mapping](#resource-mapping). A skill is always retrieved as individually addressable resources; this extension defines no packed or bundled retrieval form. See [Appendix: Deferred Features](#appendix-deferred-features).

##### Integrity and verification

> Digests are SHA-256 hashes of an artifact's raw bytes, formatted as `sha256:{hex}` where `{hex}` is 64 lowercase hexadecimal characters. A skill entry's `digest` covers its `SKILL.md` file. The index carries no digest for a skill's supporting files; they are ordinary resources, outside the integrity mechanism this extension defines.

Reviewer 8: The idea of using a digest to validate the integrity of a skill is interesting. That said, I question whether it should be a required property.

My understanding is that when a server does not provide a `skill://index.json`, the host discovers skills by using a mechanism like enumerating the resources returned by `resources/list` and selecting those with the `skill` scheme - the mechanism would be communicated via server instructions, another skill, ...

Since that discovery path does not provide a digest, I'm curious about the rationale for making the digest mandatory only for `skill://index.json`?

- An empty directory yields an empty `resources` array.
- Pagination mirrors `resources/list`: when the result includes `nextCursor`, the client passes it back as `cursor` to retrieve the next page.

A server that declares `directoryRead` MUST support the method for every directory within the skill namespaces it serves as individual files. The method itself is not skill-specific: a server MAY support it on any directory resource it serves, under any scheme.

## Implementation Guidelines

The following are recommendations for interoperable implementations. They are not part of the normative specification.

### Hosts: End-to-End Integration

This section sketches one way a host might wire MCP-served skills into an existing skills implementation. It is illustrative, not prescriptive — hosts are free to structure tools, naming, and routing however suits their architecture. The goal is that an MCP-served skill flows through the same loading and reading mechanics as a filesystem skill — while remaining origin-tagged, per [Security Implications](#security-implications).

> **Registry.** At startup and on connection change, the host assembles a single internal skill registry from every origin it supports: filesystem skill directories, and `skill://index.json` from each connected MCP server that declares the `io.modelcontextprotocol/skills` extension. Each registry entry records the skill's `name` and `description` (from the index entry's `frontmatter`) and its origin — for a filesystem skill, the local directory; for an MCP skill, the server identity and the `SKILL.md` resource URI. Because names collide within and across origins ([Names](#names)), the registry keys entries by origin and name together, qualifying colliding names for display and invocation rather than dropping either entry.

Reviewer 8: Should this paragraph describe the expected behavior for when the server provides a conventions for resources (like the `skill://` resources with server instructions) but not `skill://index.json`?

I'd expect something like

> and, from each connected MCP server that declares the `io.modelcontextprotocol/skills` extension, the skills enumerated by `skill://index.json` or, if that's not available, the resources indicated by the server via different mechanisms like  "server instructions to list all `skill://` resources.

Author: There's no normative behavior for skills referenced by just server instructions or any other non-index means, i.e. specifically the host should NOT be parsing instructions to find `skill://` resources. If a server wants to enumerate skills it should use the index.

Reviewer 8: Thank you! This clarifies that I was conflating enumeration with explicit references. Also, I asked the question before the change to use`skill/list`

| `skills[].archives[].digest`   | Yes        | SHA-256 digest of the archive file. See [Integrity](#integrity-and-verification).                                                             |

¹ Every entry MUST include `url`, a non-empty `archives`, or both.

Clients SHOULD ignore unrecognized fields. The index carries no version marker: the format is versioned by the extension itself, and an incompatible revision would be negotiated through the [capability declaration](#capability-declaration). Tolerating unrecognized fields is what allows the format to grow additively.

##### Frontmatter

`frontmatter` is the skill's `SKILL.md` YAML frontmatter rendered verbatim as a JSON object — every field the author wrote, not a curated subset. Because the Agent Skills specification requires `name` and `description`, those fields are always present; everything else (`license`, `metadata`, fields added by future revisions of the Agent Skills specification) passes through unchanged. A host can therefore build its skill registry — names, descriptions, and whatever other metadata it understands — from a single read of the index, without fetching each `SKILL.md`.

> The `frontmatter` object MUST be identical in content to the frontmatter of the `SKILL.md` it describes. When `url` is present, the final `<skill-path>` segment of the URI MUST equal `frontmatter.name`, per [Resource Mapping](#resource-mapping).

Reviewer 6: Should we reserve the `io.modelcontextprotocol/` key prefix within frontmatter `metadata` for MCP-defined semantics? These keys would be defined as part of this extension (and future MCP extensions), and only explicitly specified keys should be recognized under this prefix.

Open question: should unrecognized keys under the reserved prefix be silently ignored/stripped, or should implementations error? Ignoring is more forward-compatible (a host that predates a newly defined key doesn't break), but erroring prevents accidental misuse of the namespace.

```suggestion
The `frontmatter` object MUST be identical in content to the frontmatter of the `SKILL.md` it describes. When `url` is present, the final `<skill-path>` segment of the URI MUST equal `frontmatter.name`, per [Resource Mapping](#resource-mapping).

Within the frontmatter `metadata` object, keys prefixed with `io.modelcontextprotocol/` are reserved for semantics defined by MCP extensions. Only keys explicitly specified by a published extension are permitted under this prefix; implementations SHOULD ignore unrecognized keys in the reserved namespace. This extension currently defines no `io.modelcontextprotocol/`-prefixed metadata keys.
```

- **Inconsistent ad-hoc solutions.** Absent a convention, several independent implementations have each invented their own `skill://` URI structure, with diverging semantics for authority, path, and sub-resource addressing.

## Specification

### Dependencies

This extension has no dependencies beyond the base MCP Resources primitive.

### Skill Format

A skill served over MCP MUST conform to the [Agent Skills specification](https://agentskills.io/specification). In particular:

- A skill is a directory. Its _skill name_ is the value of the `name` field in its `SKILL.md` frontmatter.
- Every skill MUST contain a `SKILL.md` file at its root.
- `SKILL.md` MUST begin with YAML frontmatter containing at minimum the `name` and `description` fields as defined by the Agent Skills specification.
- A skill MAY contain additional files and subdirectories (references, scripts, examples, assets).

> This extension does not redefine, constrain, or extend the skill format. Future revisions of the Agent Skills specification apply automatically.

Reviewer 9: I am curious how we potentially adapt (or adjust) if there are upstream breaking changes (do we expect any with Skills)? I suspect that for the most part this is a non-issue since this spec formalizes the _distribution_ mechanism, but it might be worth spelling out.

Author: I can add more text here. Generally, I would not expect breaking changes from skills without some form of backwards compatibility otherwise everyone's skills would just stop working (which would include MCP-delivered ones).

For some clarity, could add some language like:

> In the event that the Agent Skills specification changes in a backwards incompatible way, clients MUST honor any backwards compatibility mechanisms provided by the Agent Skills specification and SHOULD continue to support the Agent Skills specification as it existed prior to any incompatible change.

Reviewer 6: We could also just advertise the supported agentskills spec range in the extension's capabilites.

Author: Unfortunately agent skills doesn't have any versioning.

- `<file-path>` is the file's path relative to the skill directory root, using `/` as the separator.

The resource for the skill's required `SKILL.md` is therefore always addressable as `skill://<skill-path>/SKILL.md`, and the skill's root directory is the URI obtained by stripping the trailing `SKILL.md`.

The final segment of `<skill-path>` MUST equal the skill's `name` as declared in its `SKILL.md` frontmatter. This mirrors the Agent Skills specification's requirement that `name` [match the parent directory name](https://agentskills.io/specification#name-field). Preceding segments, if any, are a server-chosen organizational prefix — servers MAY organize skills hierarchically by domain, team, version, or any other axis. In `skill://acme/billing/refunds/SKILL.md`, the prefix is `acme/billing` and the skill's `name` is `refunds`; in `skill://git-workflow/SKILL.md` there is no prefix and the `name` is `git-workflow`. This means the skill name is always recoverable from the URI alone, without reading frontmatter.

Further constraints:

> - A `SKILL.md` MUST NOT appear in any descendant directory of a skill. The skill directory is the boundary; skills do not nest inside other skills.

Reviewer 10: I don't understand this constraint. I think it's quite valid to have skills nested into skills. Do we have a very strong reason to impose such constraint and reduce forward compatibility?

Author: I think we can relax this. I think there was a good reason earlier on but I don't recall it and maybe the reason just went away after some iterations. Not sure if you remember any reason for this @olaservo? Seems actually quite useful to nest sub-skills for the pathological cases.

Reviewer 11: The Agent Skills spec currently does not specify how skill should be packaged or related to each other. Many clients only support skills at the top level, with a handful supporting nesting (although sometimes only a level or two).

Reviewer 12: If a skill is nested inside the folder; does that change how index.json looks? Is the recommendation that the  index.json only highlight the top level skills?

Reviewer 2: @pja-ant I think that nesting used to be more ambiguous when the URI shape was doing more of the work. The index and the directory-read method came after we added this constraint. Now that we have those, skill identity can get asserted by the index rather than inferred from the path, so I think we can relax it.

Like Kurtis mentioned, the spec doesn't disallow it and client nesting support varies. Should we frame this as servers MAY nest and not something hosts must support? A host that only handles top-level skills can still treat every entry as flat like they would in a plain filesystem.

@addy007-icy I think a nested `SKILL.md` is its own skill root with its own index entry, so index.json stays a flat list and nested skills just show up as additional entries whose URLs share a path prefix with the parent. That way a host walking the parent treats a child `SKILL.md` as a boundary rather than subsuming the entire subtree. That could keep relative-path resolution clearer and enable the top-level-skill-as-entry-point pattern that we have mentioned as a valid way to handle a giant/hierarchical set of skills.

#### Examples

| Skill path             | File                  | Resource URI                                     |
| ---------------------- | --------------------- | ------------------------------------------------ |
| `git-workflow`         | `SKILL.md`            | `skill://git-workflow/SKILL.md`                  |
| `pdf-processing`       | `references/FORMS.md` | `skill://pdf-processing/references/FORMS.md`     |
| `pdf-processing`       | `scripts/extract.py`  | `skill://pdf-processing/scripts/extract.py`      |
| `acme/billing/refunds` | `SKILL.md`            | `skill://acme/billing/refunds/SKILL.md`          |
| `acme/billing/refunds` | `examples/email.md`   | `skill://acme/billing/refunds/examples/email.md` |

#### Resource Metadata

For each `skill://<skill-path>/SKILL.md` resource:

> - `mimeType` SHOULD be `text/markdown`.

Reviewer 10: We had a brief discussion in ai-catalog about this. It's not fully clear that any form of Markdown allows for frontmatter. I *think* this is fine, but it might be strictly speaking wrong.

Author: Yeah I think strictly speaking you are right, but the agent skills spec is `SKILL.md` (markdown suffix) with YAML frontmatter. I think the ship has sailed...

- `name` SHOULD be set from the `name` field of the `SKILL.md` YAML frontmatter. By the path constraint above, this will equal the final segment of `<skill-path>`.
- `description` SHOULD be set from the `description` field of the `SKILL.md` YAML frontmatter.

> Servers MAY expose additional frontmatter fields via the resource's `_meta` object. When `_meta` keys are used for skill resources, implementations SHOULD use the `io.modelcontextprotocol.skills/` reverse-domain prefix. Other files in the skill use the `mimeType` appropriate to their content.

Reviewer 10: I wonder if we do want to be more specific here. I assume that gateway authors would really love some definitions here to ensure they can rely on governance. E.g. we can say you *must* map the frontmatter to _meta, hence allowing additional governance at intermediate steps.

I am sure you all thought about this, so I am curious what am I missing here.

Author: There's not really a lot in the skills spec frontmatter that relates to governance. If people do add their own governance metadata then of course they can add it and this line gives a carve out for where it should go, but forcing implementations to mirror it just introduces potential for things to get out of sync (servers that are fronting dynamic skills would need to expand and re-parse the frontmatter on any upstream change) and it otherwise just adds a lot of overhead for unclear value.

Reviewer 11: Generally I think the thought is that the hierarchy structure will be used for governance. E.g. Grant access to all skills under `acme/billing`. The resource name is already a header so I think it should be in the right place. 

My 2c is that we don't need to explicitly say "Servers MAY expose additional frontmatter fields" -- servers can kinda already do whatever they want in _meta today, and clients can't count on it and it should still match the SKILL.md.

### Discovery

A server is not required to make its skills enumerable. A skill's URI is directly readable via `resources/read` whether or not it appears in any index, and hosts MUST support loading a skill given only its URI (see [Hosts: End-to-End Integration](#hosts-end-to-end-integration)). This is the baseline: if a model has the URI — from server instructions, from another skill, from the user — it can read the skill.

On top of that baseline, two discovery mechanisms are defined. A server MAY support either or both.

> #### Enumeration via `skill://index.json`

Reviewer 10: One thing is that this file might become very large right? It also means it might be served without `size` so a potential unconstraint read, and lastly, since it's `json` it can't be read sequentially or partially. @sambhav here who mentioned that he might have some pathological use-cases. also @clareliguori because AWS also tends to have those. Any concerns?

Author: We discussed a bit in the working group. @jonathanhefner (IIRC) made a good point that even with 10000 skills, we're talking megabytes (uncompressed) at most. It's not a huge amount.

There are alternatives as well for servers that have far too many skills, e.g. just adding a `find_skill(query)` tool that returns a ResourceLink. Alternatively you can have skill groups where e.g. you just add a `skill://writing/SKILL.md` to your index and when you read the skill it just says "go and read skill://writing/genres to find other more specific instructions" and the model can explore with the directory read verb. Of course that wouldn't include all 10000 of your skills in the system prompt, but if you have >10000 skills then you don't want to do that anyway.

This all feels like it plays into the broader progressive discovery picture and trying to solve these pathological cases here feels like trying to boil the ocean.

Reviewer 6: I'm fine with it not being paginated, even with 100k entries it will still be reasonable to fetch. Also the SEP specifies that not every skill needs to be in the index, which is perfect as it allows for progressive discovery if needed.

One option if we are really worried is making it jsonl, but this is fine for now.

> A server SHOULD expose a resource at the well-known URI `skill://index.json` whose content is a JSON index of the skills it serves.

Reviewer 10: Hm. This makes sense. One thing I Worry about is that Gateway authors cannot easily introspect this on the protocol level or enhance attributes via _meta, e.g. signing them or verifying that a skill has been checked/verified in a form, or add other provenance or policy metadata. I wish we could find a way to put this into _meta on a directory/read, or a new skills/list. I am a bit torn, I do think modelling it purely on resources is a good idea in principle for backwards compatibility but it comes with some real drawbacks.

Author: I mean you can put `_meta` on a `resources/read` of `skill://index.json` and the index contains the full frontmatter which is another place to put that kind of metadata that is specific to skills. The `resources/read` on the individual skills can also contains `_meta` metadata.

```json
{
  "skills": [
    {
      "url": "skill://git-workflow/SKILL.md",
      "digest": "sha256:a1b2c3d4...",
      "frontmatter": {
        "name": "git-workflow",
        "description": "Follow this team's Git conventions for branching and commits"
      }
    },
    {
      "url": "skill://acme/billing/refunds/SKILL.md",
      "digest": "sha256:b2c3d4e5...",
      "frontmatter": {
        "name": "refunds",
        "description": "Process customer refund requests per company policy",
        "license": "Apache-2.0"
      },
      "archives": [
        {
          "url": "skill://acme/billing/refunds.tar.gz",
          "mimeType": "application/gzip",
          "digest": "sha256:c4d5e6f7..."
        }
      ]
    },
    {
      "frontmatter": {

>         "name": "pdf-processing",

Reviewer 9: This is mostly a nit to make this SEP easier to parse-out - everywhere else in the doc you're referring to the contents of the `pdf-processing` skill as a non-archive one. Just a cohesiveness bit that will reduce confusion.

##### Archives

`archives` lists pre-packed forms of the entire skill directory. Each archive is a single resource whose content is the complete skill directory in an archive format; reading it via `resources/read` retrieves the skill — `SKILL.md` and all supporting files — in one round trip. When multiple archives are listed, they are alternative encodings of identical content; a host picks one whose format it supports, determined from `mimeType`.

`mimeType` MUST identify an archive format. Hosts SHOULD support at least gzip-compressed tar (`application/gzip`, conventionally `.tar.gz`) and ZIP (`application/zip`); servers MAY offer additional formats but SHOULD include at least one of these two so that any conforming host can retrieve the skill.

Archive layout:

- `SKILL.md` MUST be at the archive root, not nested inside a wrapper directory.
- Entry paths MUST use `/` as the separator and MUST be relative — no path-traversal sequences (`..`) and no absolute paths.

> Archive contents represent the skill directory directly. When an entry carries both `url` and `archives`, each archive MUST unpack to the same files that are individually addressable under the skill path — a host may fetch the skill either way and observes the same content. When an entry is archive-only, the skill's files are not individually addressable on the server; the host unpacks the archive locally and addresses its contents under `skill://<name>/<file-path>`, where `<name>` is `frontmatter.name`. In both cases the archive `url` itself carries no path semantics — it is a download location, nothing more.

Reviewer 9: This feels a bit counter to the previous guarantees of skill readability:

> A server is not required to make its skills enumerable. A skill's URI is directly readable via `resources/read` whether or not it appears in any index, and hosts MUST support loading a skill given only its URI (see [Hosts: End-to-End Integration](#hosts-end-to-end-integration)). This is the baseline: if a model has the URI — from server instructions, from another skill, from the user — it can read the skill.

Wouldn't this effectively contradict the `MUST` you just mentioned about addressability of a skill for an entire distribution class?

This also might introduce weirdness because the client now needs to know that the skill came from the archive vs. server and route to the unpacked version. Should there be a carve-out for how clients can address a local extracted cache vs. server-served skills?

Author: Not sure I follow. For archives, the skill will still be available via the archive, e.g. a `skill://code-review.zip` is still readable via `read_resource` even if that skill is not in the index.

Author: I could clarify with:

> A skill's URI — either its `SKILL.md` resource or an archive resource containing it — is directly readable via `resources/read` whether or not it appears in any index…

> Archive contents represent the skill directory directly. When an entry carries both `url` and `archives`, each archive MUST unpack to the same files that are individually addressable under the skill path — a host may fetch the skill either way and observes the same content. When an entry is archive-only, the skill's files are not individually addressable on the server; the host unpacks the archive locally and addresses its contents under `skill://<name>/<file-path>`, where `<name>` is `frontmatter.name`. In both cases the archive `url` itself carries no path semantics — it is a download location, nothing more.

Reviewer 9: Another point here:

> "the host unpacks the archive locally and addresses its contents under `skill://<name>/<file-path>`, where `<name>` is `frontmatter.name`"

`frontmatter.name` is `refunds`. So `acme/billing/refunds` unpacks to `skill://refunds/...`. And `acme/support/refunds` also unpacks to `skill://refunds/...`. The two skills the prefix existed to separate now address the same location on the host. Second one fetched overwrites the first, or the host has two skills at one URI, depending on implementation.

Do you need to introduce a `skillPath`?

Author: urgh, yeah this section is incorrect. The archive should unpack to just raw reference paths, not resource URIs (since you aren't going to load these as resources anyway).

Should be something like:

> …the host unpacks the archive locally; relative references within the skill resolve against the unpacked root, per Internal References.

Compared to individual-file distribution, an archive delivers a multi-file skill atomically and can carry UNIX file metadata (executable bits, symlinks) that individual resources cannot represent — robustly for `.tar.gz`, variably for `.zip`.

> Archives are an unpacking attack surface. Hosts unpacking an archive MUST reject archives containing path-traversal sequences or absolute paths, reject symlinks or hard links that resolve outside the skill directory, and enforce a limit on total unpacked size to prevent decompression bombs.

Reviewer 10: What's the motivation for including archives? This feels like something that clients will have a hard time reasoning about, since there is no additional IT admin control if the archives don't contain zip-bombs, malicious materials, etc, compared to a marketplace.

Author: Archives are highly requested here since otherwise a skill that potentially contains 100s or 1000s of files would require that the client pulls them one-by-one using `resources/read`, putting a lot of unnecessary strain on both the server and client. They are also a standard format for distributing skills.

It's true that zip bombs are a threat, but you have to deal with that anyway: even if the skill wasn't archived for delivery, it could still have a reference file `skill://my-skill/references/data.zip` that the client has to extract. Clients just need to deal with this.

> Archives are an unpacking attack surface. Hosts unpacking an archive MUST reject archives containing path-traversal sequences or absolute paths, reject symlinks or hard links that resolve outside the skill directory, and enforce a limit on total unpacked size to prevent decompression bombs.

Reviewer 9: I think the security considerations here are underspecified. We're basically giving implementers a footgun and hoping for the best.

Reviewer 9: ```suggestion
Archives are an unpacking attack surface. Hosts unpacking an archive MUST:

1. Reject entries with path-traversal sequences or absolute paths.
2. Reject symlinks or hard links that resolve outside the skill directory.
3. Extract only regular files, directories, and symlinks, rejecting device nodes and any other special-file entry type.
4. Reject archives in which any two entry paths are equal after Unicode normalization and case folding, as a case/normalization collision can silently overwrite `SKILL.md` on case-insensitive or normalizing filesystems.
5. Clear setuid/setgid/sticky mode bits on every extracted entry and extract all entries as the host process's own user and group, ignoring uid/gid fields recorded in the archive.
6. Enforce limits on total unpacked size, total entry count, and maximum path depth, applied both per archive and as a cumulative per-server budget. Once the per-server budget is exhausted, hosts MUST refuse to unpack further archives from that server.

Hosts SHOULD use existing hardened libraries for handling archives, specific to their language and platform rather than reimplement them.

Digest verification ([Integrity and verification](#integrity-and-verification)) covers the archive bytes at retrieval only, not the unpacked tree thereafter. Hosts MUST unpack to a host-private location that is not within, and not writable via, any filesystem path exposed to MCP tools or to the model's execution environment, and MUST keep the unpacked tree immutable for the cached skill's lifetime.
```

##### Integrity and verification

> Digests are SHA-256 hashes of an artifact's raw bytes, formatted as `sha256:{hex}` where `{hex}` is 64 lowercase hexadecimal characters. For a skill entry's `digest`, the artifact is the `SKILL.md` file; for an archive's `digest`, the artifact is the archive file.

Reviewer 10: Why use hexadecimal and not BASE64? Most modern SRI formats use base64. We might want to do that? We might even just use SRI so people can even choose if they want sha256,sha384 or sha512. It's the de-factor default for modern package management / integrity verification and serializes more compact.

https://developer.mozilla.org/en-US/docs/Web/Security/Defenses/Subresource_Integrity

Author: Honest answer is that we were following https://github.com/cloudflare/agent-skills-discovery-rfc#integrity-and-verification initially.

Other answer: hex appears to be the industry standard here. base64 seems to be a primarily an npm-ecosystem thing, but outside of that hex dominates: [OCI](https://github.com/opencontainers/image-spec/blob/main/descriptor.md#digests), [cargo](https://doc.rust-lang.org/cargo/reference/registry-index.html?highlight=cksum#json-schema), [pip](https://peps.python.org/pep-0503/#specification), [homebrew](https://github.com/mapbox/homebrew/blob/master/share/doc/homebrew/Formula-Cookbook.md#grab-the-url), [apt](https://wiki.debian.org/DebianRepository/Format#:~:text=apt/apt_0.9.3_amd64.deb-,Size%2C%20MD5sum%2C%20SHA1%2C%20SHA256%2C%20SHA512,-The%20mandatory%20Size) - all use hex.

Reviewer 6: This is correct, the hex digest is the prevelant norm for this.

> Digests are SHA-256 hashes of an artifact's raw bytes, formatted as `sha256:{hex}` where `{hex}` is 64 lowercase hexadecimal characters. For a skill entry's `digest`, the artifact is the `SKILL.md` file; for an archive's `digest`, the artifact is the archive file.

Reviewer 9: I am a bit uneasy about the digest logic, TBH. We're saying that the artifact is `SKILL.md` but then there is also the ability to load supporting files, like scripts. If the skill is "Use `server.py` to do FooBar" and we only create a digest for the Markdown file, doesn't it mean that now any of the supporting code can be... whatever server wants it to be?

Author: The digest is primarily for caching purposes to avoid refetching the skill.

For verification, I think at some level you have to accept that you can't lock down absolutely everything. If we (somehow) expanded the digest to cover all the relevant resources, but then the skill says something like "call tool `get_script`, which returns a script file and run that" - we can't guarantee that the tool result is consistent across every usage.

> Hosts MUST verify retrieved content against the `digest` in the index. A mismatch indicates the content is corrupted or tampered with — hosts MUST NOT use unverified content. Digests also enable efficient caching: compare a skill's `digest` against a locally cached value to determine whether the artifact has changed without re-reading it.

Reviewer 10: One thing to consider is to enforce `size` being available for any served file for skill. I think we missed this in the original spec, but now that you enforce verification, you enforce that the client reads the full content and hence it needs to understand the max bounds it should read in advance or discard if it's too big.

So maybe we say, if you have a skill you MUST use `size`

Author: Makes sense. Can add it.

> 

Reviewer 9: ```suggestion

Hosts MUST verify retrieved content against the `digest` in the index. A mismatch indicates the content is corrupted or tampered with - hosts MUST NOT use unverified content. Digests also enable efficient caching: compare a skill's `digest` against a locally cached value to determine whether the artifact has changed without re-reading it.

After fetching and digest-verifying a `SKILL.md` whose entry appeared in the index, hosts MUST parse its YAML frontmatter and compare it field-by-field against the entry's `frontmatter`. Any discrepancy MUST be treated as a verification failure equivalent to a digest mismatch, and the skill MUST NOT be loaded. This enforces the [Frontmatter](#frontmatter) identity requirement on the host side, so that what a user approves from the index is what the model actually receives.

A host that retrieves a skill via an archive and verifies the archive digest MUST thereafter serve every read of that skill's files from the locally unpacked archive for the lifetime of that verification. It MUST NOT satisfy those reads by issuing live `resources/read` calls against the server, because the archive digest is the only integrity check covering supporting files. A malicious server can return different bytes live than the verified archive contained.
```

Internal references within a skill (e.g., `SKILL.md` linking to `references/GUIDE.md`) are relative paths, as in the filesystem form of the Agent Skills specification. A client resolves a relative reference against the skill's root — `references/GUIDE.md` in `skill://acme/billing/refunds/SKILL.md` resolves to `skill://acme/billing/refunds/references/GUIDE.md` — exactly as a filesystem path would resolve. The skill's root is the directory containing `SKILL.md`, not the `skill://` scheme root.

### Directory Listing

A skill's instructions frequently reference a directory rather than a file: "pick the appropriate template from `templates/`", "run the matching script in `scripts/`". To act on this, the agent must learn what the directory contains. `resources/list` cannot answer that scoped question: it enumerates the server's entire resource space, not a subtree, and the servers this SEP most wants to accommodate — large, generated, or unenumerable catalogs (see [Why Is Enumeration Optional?](#why-is-enumeration-optional)) — may not implement meaningful global listing at all.

> This extension therefore defines one new method, `resources/directory/read`, gated behind the `directoryRead` setting of the [capability declaration](#capability-declaration).

Reviewer 10: We should give this lots of scrutiny. I assume that this will be the basis for a rework of resource handling to look more like a filesystem API.

#### Directory resources

> 

Reviewer 10: Effectively this replace `resource/list` with a new thing model after `readdir`?

Author: It doesn't replace `resources/list` (which must exist for backwards compatibility) but it does solve for the `readdir` use case where the model wants to explore a sub directory of a skill without having to list every single resource on the server.

Perhaps in the future we could deprecate `resources/list` and just have the server expose some roots that you can start using this new primitive on. That would be a good discussion for later, but for this extension we only need the `readdir` behavior and want to keep the scope minimal to avoid a potentially lengthy and deep design discussion.

> A _directory resource_ is a resource whose `mimeType` is `inode/directory`. In a skill namespace served as individual files, every directory level is a directory resource: the skill root (`skill://pdf-processing`) and each subdirectory (`skill://pdf-processing/templates`). Directory URIs are written without a trailing slash. Directory resources need not appear in `resources/list`; they are addressable whether listed or not.

Reviewer 10: Why does the URI need to be without a trailiing slash? that feels like an arbitrary limitation?

Author: The alternative of allowing either IMO would just lead to incompatibility: some servers may only accept one or the other / some clients may normalize one way or the other -- leads to incompatibility. Having one correct approach and avoids the need for normalization (e.g. you blocklist access to `skill://dangerous/dir` but the client requests `skill://dangerous/dir/` and if you forget to normalize then it falls over).

Existing implementations using other `skill://` URI structures will need to adjust to conform — see the Working Group's [related-work survey](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/related-work.md) for a catalog. Notably, FastMCP's widely-used [SkillsProvider](https://gofastmcp.com/servers/providers/skills) diverges on URI structure, discovery (per-skill `_manifest` vs. central index), and metadata mapping; coordinating that migration is a near-term Working Group priority. These are mechanical changes, not semantic ones.

## Security Implications

Skill content is instructional text delivered to a model, which makes it a prompt-injection surface. The Working Group's position, recorded in [open-questions.md §10](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/open-questions.md#10-how-should-skills-handle-security-and-trust-boundaries), is:

> - **Skill content is untrusted input.** Hosts MUST treat MCP-served skill content as untrusted model input, subject to the same prompt-injection defenses applied to any server-provided text. A server being connected does not make its skill content authoritative.

Reviewer 9: ```suggestion
- **Skill content is untrusted input.** Hosts MUST treat MCP-served skill content as untrusted model input, subject to the same prompt-injection defenses applied to any server-provided text. A server being connected does not make its skill content authoritative.
- **Origin MUST be visible to the model.** Hosts MUST tag MCP-served skill content with its originating server identity at the point it enters model context and MUST NOT present an MCP-served skill to the model as indistinguishable from a local filesystem skill. The model, not the host, decides whether to follow a skill's instructions. Withholding origin from it makes the untrusted-input requirement above unenforceable at the layer that acts on it.
```

> - **Skills do not introduce a new trust tier.** A user who connects a server has already extended their trust boundary to it; a malicious server can do as much harm via tools as via a skill document. Serving skills over MCP adds no risk beyond what skills already carry in any transport — but the defensive posture above applies regardless.

Reviewer 9: I would actually disagree here. A single server (e.g., remote server) has a limited blast radius with its tools since they do not offer local execution. A skill with additional files does just that - suddenly, I now can have code running on my machine that did not happen before. This need very explicit clarification.

Reviewer 9: ```suggestion
- **Skills introduce host-side surfaces that tools do not.** Unlike a remote tool call, an MCP-served skill can place server-authored bytes on the host filesystem direct the model to execute them with host-side tools. Hosts MUST treat MCP-served skills as a higher-risk surface than remote tool invocation.
```

> - **No implicit local execution.** Hosts MUST NOT honor mechanisms in skill content that would cause local code execution without explicit user opt-in. This includes, non-exhaustively: hook declarations, pre/post-invocation scripts, shell commands embedded in frontmatter, or any field that a filesystem-sourced skill might use to register executable behavior on the host. Hosts MUST either ignore such fields entirely when the skill arrives over MCP, or gate them behind an explicit per-skill user approval that states what will execute and where. Silently executing server-provided code because it appeared in a skill directory is a remote code execution vector.

Reviewer 9: ```suggestion
- **No implicit local execution.** Hosts MUST NOT allow MCP-served skill content to cause host-side code execution without explicit per-skill user approval. This covers (a) declarative fields the host parses (hooks, frontmatter scripts) and (b) instructions in the skill body that direct the model to invoke any host code-execution tool, whether to run a script bundled in the skill or to run an arbitrary command the skill specifies. Hosts MUST ignore or approval-gate (a), and MUST apply the same approval gate to code-execution tool calls issued while the model is acting on an MCP-served skill.
- **Origin-scoped resource reads.** A model-callable resource-read surface (such as the `read_resource` pattern in [Hosts: End-to-End Integration](#hosts-end-to-end-integration)) is a cross-server confused-deputy vector when driven by untrusted skill content. Hosts MUST bind such reads to the skill's originating server: a skill served by server A MUST NOT cause a `resources/read` against server B. Hosts MUST identify servers by a host-assigned label, not the server's self-reported `serverInfo.name`. Any cross-origin read MUST be gated behind explicit per-call user approval naming both servers.
```

> - **Skills are data, not directives.** Hosts MUST NOT treat skill resources as higher-authority than other context. Explicit user policy governs whether a skill is loaded at all.

Reviewer 9: ```suggestion
- **No implicit permission grants.** Hosts MUST NOT honor frontmatter fields that widen the model's tool or filesystem permissions when the skill arrives over MCP. In particular, the Agent Skills `allowed-tools` field, which a filesystem-sourced skill uses to declare the tools available while it runs, MUST be ignored for MCP-origin skills unless the user has explicitly approved that grant for that skill. A remote server populating `allowed-tools` is requesting elevated access on the host, not declaring a property of its own environment.
- **Skills are data, not directives.** Hosts MUST NOT treat skill resources as higher-authority than other context. Explicit user policy governs whether a skill is loaded at all.
```

> - **Provenance and inspection.** Hosts SHOULD indicate which server a skill originates from when presenting it, SHOULD let users inspect a skill's content before it is loaded into model context, and MAY gate loading behind per-skill or per-server user approval.

Reviewer 9: These need to be MUSTs. From a security standpoint, hosts should never let skills transit from server to host without transparent origin and consent, because that now enables local execution.

Author: Generally agree for standard chat-bot like clients, but I think it depends on context. e.g. for skills that have already been vetted in some way or pre-signed off on (particularly thinking about fully autonomous use cases) then clients would need a way to use those skills without a human in the loop.

For example: a customer support agent may use a skill to guide their conversation with customers. It would be a show-stopper if that skill needed a human sign off before the agent could do its job.

- **Archive safety.** Archive-distributed skills are an unpacking attack surface. Hosts MUST validate archives per the requirements in [Archives](#archives): reject path traversal and absolute paths, reject links resolving outside the skill directory, and bound total unpacked size to prevent decompression bombs.

> 

Reviewer 9: ```suggestion
- **Content-bound approval.** When a host persists any per-skill user approval, it MUST be bound to the `digest` of the `SKILL.md` (and the archive `digest`, if present) observed at the moment of approval. If a subsequent index advertises a different digest under the same skill name or URI, the host MUST treat the prior approval as revoked and re-prompt before loading or executing. Digest verification ([Integrity and verification](#integrity-and-verification)) defends only against transport tampering - it cannot defend against the originating server rotating content, because the server authors both the index and the body.
- **Cache isolation and durable origin.** Hosts that unpack or cache MCP-served skill content on disk MUST do so in a location excluded from every filesystem-skill discovery path, and MUST treat content loaded from that location as having arrived over MCP for all purposes of the no-implicit-local-execution rule above, including after host restart and after the originating server is disconnected. Unpacked bytes do not graduate to filesystem-skill trust by residing locally. Hosts SHOULD remove a server's cached skill content when the user removes that server.

```

Hosts SHOULD use existing hardened libraries for handling archives, specific to their language and platform rather than reimplement them.

Digest verification ([Integrity and verification](#integrity-and-verification)) covers the archive bytes at retrieval only, not the unpacked tree thereafter. Hosts MUST unpack to a host-private location that is not within, and not writable via, any filesystem path exposed to MCP tools or to the model's execution environment, and MUST keep the unpacked tree immutable for the cached skill's lifetime.

##### Integrity and verification

Digests are SHA-256 hashes of an artifact's raw bytes, formatted as `sha256:{hex}` where `{hex}` is 64 lowercase hexadecimal characters. For a skill entry's `digest`, the artifact is the `SKILL.md` file; for an archive's `digest`, the artifact is the archive file.

Hosts MUST verify retrieved content against the `digest` in the index. A mismatch indicates the content is corrupted or tampered with — hosts MUST NOT use unverified content. Digests also enable efficient caching: compare a skill's `digest` against a locally cached value to determine whether the artifact has changed without re-reading it.

> Hosts MUST verify retrieved content against the `digest` in the index. A mismatch indicates the content is corrupted or tampered with - hosts MUST NOT use unverified content. Digests also enable efficient caching: compare a skill's `digest` against a locally cached value to determine whether the artifact has changed without re-reading it.

Reviewer 9: This line duplicates L215. Replacing it with the not-a-security-boundary clarification:
```suggestion
Digests are unsigned and supplied by the same server that supplies the content. A match proves the two are consistent, not that either is trustworthy. Any intermediary on the path, such as a gateway, can rewrite both the index and the content together. Hosts MUST NOT treat a digest match as a security boundary.
```

- **No implicit permission grants.** Hosts MUST NOT honor frontmatter fields that widen the model's tool or filesystem permissions when the skill arrives over MCP. In particular, the Agent Skills `allowed-tools` field, which a filesystem-sourced skill uses to declare the tools available while it runs, MUST be ignored for MCP-origin skills unless the user has explicitly approved that grant for that skill. A remote server populating `allowed-tools` is requesting elevated access on the host, not declaring a property of its own environment.
- **Skills are data, not directives.** Hosts MUST NOT treat skill resources as higher-authority than other context. Explicit user policy governs whether a skill is loaded at all.
- **Provenance and inspection.** Hosts SHOULD indicate which server a skill originates from when presenting it, SHOULD let users inspect a skill's content before it is loaded into model context, and MAY gate loading behind per-skill or per-server user approval.

> - **Archive safety.** Archive-distributed skills are an unpacking attack surface. Hosts MUST validate archives per the requirements in [Archives](#archives): reject path traversal and absolute paths, reject links resolving outside the skill directory, and bound total unpacked size to prevent decompression bombs.

Reviewer 9: ```suggestion
- **Archive safety.** Archive-distributed skills are an unpacking attack surface. Hosts MUST validate archives per the requirements in [Archives](#archives): reject path traversal and absolute paths, reject links resolving outside the skill directory, and bound total unpacked size to prevent decompression bombs.
- **Digests are not a security boundary.** Index digests are unsigned and come from the same server as the content. They confirm consistency between the index and what was fetched, as described in [Integrity and verification](#integrity-and-verification), but they cannot establish trust in the content, defend against the server itself, or detect an intermediary that rewrites both together.
```

| `skills[].resources[].uri`    | Yes      | Resource URI of the file.                                                                                    |
| `skills[].resources[].digest` | Yes      | SHA-256 digest of the file. See [Integrity](#integrity-and-verification).                                    |

¹ MAY be omitted only for dynamically generated skills whose content cannot be pre-digested. Hosts MAY decline to load skills without `resources`.

Pagination mirrors the base protocol's list methods: the request accepts an optional `cursor`, and when the result includes `nextCursor` the client passes it back to retrieve the next page. An entry is atomic — a skill's `resources` set is never split across pages.

In protocol versions 2026-07-28 and later, the result also carries the base protocol's list-caching attributes — `ttlMs` and `cacheScope`, as defined for `tools/list` and `resources/list` ([SEP-2549]) — with the same semantics: a freshness hint for the listing and a cache-scope marker, not an integrity property. [Integrity and verification](#integrity-and-verification) governs content regardless of how fresh a cached listing is.

> A server whose skill catalog is large, generated on demand, or otherwise unenumerable MAY return an empty or partial listing. Hosts MUST NOT treat an empty or partial listing as proof that a server has no skills. The method serves entries for a server's skills whatever URI scheme they use — enumeration is uniform across schemes.

Reviewer 13: Given a server may return partial listing; I think we should also introduce a `skills/get` so that we can get the resource's manifest (so that the digest can be verified)  for a given skill. Otherwise there's no way to do fetch it. 

wdyt?

Reviewer 13: Addressed in the next commit

Reviewer 13: Ty for the change @pja-ant - feel free to resolve the comment. I can't.

| `skills[].resources[].uri`    | Yes      | Resource URI of the file.                                                                                    |
| `skills[].resources[].digest` | Yes      | SHA-256 digest of the file. See [Integrity](#integrity-and-verification).                                    |

¹ MAY be omitted only for dynamically generated skills whose content cannot be pre-digested. Hosts MAY decline to load skills without `resources`.

Pagination mirrors the base protocol's list methods: the request accepts an optional `cursor`, and when the result includes `nextCursor` the client passes it back to retrieve the next page. An entry is atomic — a skill's `resources` set is never split across pages.

In protocol versions 2026-07-28 and later, the result also carries the base protocol's list-caching attributes — `ttlMs` and `cacheScope`, as defined for `tools/list` and `resources/list` ([SEP-2549]) — with the same semantics: a freshness hint for the listing and a cache-scope marker, not an integrity property. [Integrity and verification](#integrity-and-verification) governs content regardless of how fresh a cached listing is.

> A server whose skill catalog is large, generated on demand, or otherwise unenumerable MAY return an empty or partial listing. Hosts MUST NOT treat an empty or partial listing as proof that a server has no skills. The method serves entries for a server's skills whatever URI scheme they use — enumeration is uniform across schemes.

Reviewer 14: `skills/list` supports pagination. For partial results response should include `nextCursor` field. This would be a partial result indicator. Would this work?

## Motivation

Native skills support in host applications demonstrates strong demand for rich, progressively disclosed workflow instructions. MCP does not currently offer a conventional way to ship this content alongside the tools it describes, which leads to:

- **Fragmented distribution.** A server and the skill that teaches an agent to use it are versioned, discovered, and installed separately. Users installing a server from a registry have no signal that a companion skill exists. ([problem statement](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/problem-statement.md))

> - **Instruction size limits.** Server instructions load once at initialization and are practically bounded in size. Complex workflows — such as the 875-line [mcpGraph skill](https://github.com/TeamSparkAI/mcpGraph/blob/main/skills/mcpgraphtoolkit/SKILL.md) — do not fit this model. ([experimental findings](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/experimental-findings.md#mcpgraph-skills-in-mcp-server-repo))

Reviewer 9: There's no initialization step anymore under 2026-07-28; instructions come back as the `instructions` field of the `server/discover` result.

```suggestion
- **Instruction size limits.** Server instructions are delivered as the `instructions` field of the `server/discover` result and are practically bounded in size. Complex workflows — such as the 875-line [mcpGraph skill](https://github.com/TeamSparkAI/mcpGraph/blob/main/skills/mcpgraphtoolkit/SKILL.md) — do not fit this model. ([experimental findings](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/experimental-findings.md#mcpgraph-skills-in-mcp-server-repo))
```

Reviewer 13: What do you think about suggesting both to keep the SEP backward compatible?

Reviewer 2: @aditya-scio for this one I think it makes sense to refer to the current state of server instructions in the latest spec, since it isn't describing the SEP itself.

- A skill is a directory. Its _skill name_ is the value of the `name` field in its `SKILL.md` frontmatter.
- Every skill MUST contain a `SKILL.md` file at its root.
- `SKILL.md` MUST begin with YAML frontmatter containing at minimum the `name` and `description` fields as defined by the Agent Skills specification.
- A skill MAY contain additional files and subdirectories (references, scripts, examples, assets).

This extension does not redefine, constrain, or extend the skill format. Future revisions of the Agent Skills specification apply automatically.

### Resource Mapping

Each file within a skill directory is exposed as an MCP resource. Servers SHOULD use the `skill://` URI scheme, under which the resource URI has the form:

```
skill://<skill-path>/<file-path>
```

where:

- `<skill-path>` is a `/`-separated path of one or more segments locating the skill directory within the server's skill namespace. It MAY be a single segment (`git-workflow`) or nested to arbitrary depth (`acme/billing/refunds`).
- `<file-path>` is the file's path relative to the skill directory root, using `/` as the separator.

> The resource for the skill's required `SKILL.md` is therefore always addressable as `skill://<skill-path>/SKILL.md`, and the skill's root directory is the URI obtained by stripping the trailing `SKILL.md`.

Reviewer 9: Stripping the literal `SKILL.md` leaves a trailing slash, but Directory Listing writes the root `skill://pdf-processing` and says directory URIs have no trailing slash. One canonical form:

```suggestion
The resource for the skill's required `SKILL.md` is therefore always addressable as `skill://<skill-path>/SKILL.md`, and the skill's root directory is `skill://<skill-path>` (the `/SKILL.md` suffix removed, no trailing slash), matching [Directory Listing](#directory-listing).
```

The final segment of `<skill-path>` MUST equal the skill's `name` as declared in its `SKILL.md` frontmatter. This mirrors the Agent Skills specification's requirement that `name` [match the parent directory name](https://agentskills.io/specification#name-field). Preceding segments, if any, are a server-chosen organizational prefix — servers MAY organize skills hierarchically by domain, team, version, or any other axis. In `skill://acme/billing/refunds/SKILL.md`, the prefix is `acme/billing` and the skill's `name` is `refunds`; in `skill://git-workflow/SKILL.md` there is no prefix and the `name` is `git-workflow`. This means the skill name is always recoverable from the URI alone, without reading frontmatter.

Further constraints:

- A `SKILL.md` MAY appear in a descendant directory of a skill — skills can nest. See [Nested skills](#nested-skills).

> - The final `<skill-path>` segment, being the skill `name`, MUST satisfy the Agent Skills specification's naming rules. Prefix segments SHOULD be valid URI path segments per [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986); no further constraints are imposed on them.

Reviewer 9: The next line says the first segment occupies the authority component, and the authority uses RFC 3986's `reg-name` grammar, which excludes the `:` and `@` that path `pchar` allows. So "valid path segments" is the wrong grammar for the first segment:

```suggestion
- The final `<skill-path>` segment, being the skill `name`, MUST satisfy the Agent Skills specification's naming rules. The first `<skill-path>` segment occupies the authority component and SHOULD be a valid `reg-name` per [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986); any other prefix segments SHOULD be valid URI path segments; no further constraints are imposed on them.
```

After fetching a `SKILL.md` for which the host holds an entry, from either `skills/list` or `skills/get` — digest-verified when the entry carries `resources`, and unverifiable when it does not — hosts MUST parse its YAML frontmatter and compare it field-by-field against the entry's `frontmatter`. Any discrepancy MUST be treated as a verification failure equivalent to a digest mismatch, and the skill MUST NOT be loaded. This enforces the [Frontmatter](#frontmatter) identity requirement on the host side, so that what a user approves from the listing is what the model actually receives.

#### Retrieval via `skills/get`

A server declaring the `io.modelcontextprotocol/skills` extension MUST also implement the `skills/get` method, which returns the entry for a single skill named by its URI:

```json
{
  "jsonrpc": "2.0",
  "id": 5,
  "method": "skills/get",
  "params": {
    "uri": "skill://pdf-processing/SKILL.md"

>   }
> }
> ```
> 
> ```json
> {
>   "jsonrpc": "2.0",
>   "id": 5,
>   "result": {
>     "skill": {
>       "uri": "skill://pdf-processing/SKILL.md",
>       "frontmatter": {
>         "name": "pdf-processing",
>         "description": "Extract, fill, and assemble PDF documents",
>         "metadata": { "version": "2.1.0" }
>       },
>       "resources": [
>         {
>           "uri": "skill://pdf-processing/SKILL.md",
>           "digest": "sha256:d5e6f7a8..."
>         },
>         {
>           "uri": "skill://pdf-processing/references/FORMS.md",
>           "digest": "sha256:e6f7a8b9..."
>         }
>       ]

Reviewer 9: `resources` is normatively complete (a read of an unlisted file is a verification failure), but this example lists only two files while the `skills/list` example shows six for the same skill with the same digests. Matching them:

```suggestion
      "resources": [
        {
          "uri": "skill://pdf-processing/SKILL.md",
          "digest": "sha256:d5e6f7a8..."
        },
        {
          "uri": "skill://pdf-processing/references/FORMS.md",
          "digest": "sha256:e6f7a8b9..."
        },
        {
          "uri": "skill://pdf-processing/scripts/extract.py",
          "digest": "sha256:f7a8b9c0..."
        },
        {
          "uri": "skill://pdf-processing/templates/invoice.md",
          "digest": "sha256:a8b9c0d1..."
        },
        {
          "uri": "skill://pdf-processing/templates/purchase-order.md",
          "digest": "sha256:b9c0d1e2..."
        },
        {
          "uri": "skill://pdf-processing/templates/regional/eu-invoice.md",
          "digest": "sha256:c0d1e2f3..."
        }
      ]
```

    }
  }
}
```

`params.uri` MUST be the URI of a skill's `SKILL.md`. The `skill` object is a skill entry, identical in shape and meaning to an entry of `skills/list` — the same `uri`, `frontmatter`, and `resources` fields, under the same rules.

Semantics:

- If the URI does not identify a skill the server serves, the server MUST return error `-32602` (Invalid params) — the same code `resources/read` uses for unknown resources.
- A server MUST answer for every skill it serves, whether or not that skill appears in its `skills/list` result. A skill absent from a partial listing is still retrievable by URI.
- The result is a point-in-time snapshot, exactly as a listing entry is. Re-calling the method is how a host refreshes one skill's digests without re-enumerating the catalog.
- A skill whose content is generated dynamically omits `resources`, per [Resources](#resources), whether it is reached through `skills/list` or `skills/get`.

> - The result carries no pagination cursor and no list-caching attributes: a single entry is not a list, and the entry is a snapshot of the skill as the server holds it at that moment.

Reviewer 9: Caching attributes aren't a list property in 2026-07-28: `resources/read` is a single-item result and it's a `CacheableResult`. Worth dropping that reasoning; whether `skills/get` itself should be cacheable is a separate call.

```suggestion
- The result carries no pagination cursor: a single entry is not a list. The entry is a snapshot of the skill as the server holds it at that moment; whether the result should also carry the base protocol's caching attributes (`ttlMs` and `cacheScope`, per [SEP-2549]), as `resources/read` results do, is left open.
```

Reviewer 9: @pja-ant @dsp-ant I wonder if we should be opinionated here, and whether we should have a strong stance on caching - not a big fan of leaving this open to interpretation.

def refunds():
    return Path("./skills/refunds")
```

The SDK handles: reading `SKILL.md` frontmatter to populate resource metadata, serving file content on `resources/read`, and answering `skills/get` — and, where the server's skill set is bounded, `skills/list` — computing entry digests from the registered files.

**Client-side** — enumerate and fetch skills:

```python
skills = await client.list_skills()               # calls skills/list, paginating; may be empty
entry = await client.get_skill(
    "skill://acme/billing/refunds/SKILL.md")      # calls skills/get, listed or not
content = await client.read_skill_uri(
    "skill://acme/billing/refunds/SKILL.md")      # wraps resources/read, works regardless of enumeration
entries = await client.read_directory(
    "skill://pdf-processing/templates")           # wraps resources/directory/read
```

These wrappers are thin — each is a single underlying protocol call with a fixed URI pattern — but they give server authors an ergonomic way to declare skills and give client authors a discoverable entry point.

## Rationale

### Why Resources Instead of a New Primitive?

> The Working Group's [decision log](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/decisions.md#2026-02-26-prioritize-skills-as-resources-with-client-helper-tools) records this as settled. Skills are files; Resources exist to expose files. Reusing Resources inherits URI addressability, `resources/read`, `resources/subscribe`, and the existing client tooling for free. A new primitive would duplicate most of this and add ecosystem complexity. Using resources to describe files is also aligned with [Composability over specificity](https://modelcontextprotocol.io/community/design-principles#composability-over-specificity) and other MCP [design principles](https://modelcontextprotocol.io/community/design-principles).

Reviewer 9: `resources/subscribe` was removed in 2026-07-28; the equivalent is now the `resourceSubscriptions` filter on `subscriptions/listen`.

```suggestion
The Working Group's [decision log](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/decisions.md#2026-02-26-prioritize-skills-as-resources-with-client-helper-tools) records this as settled. Skills are files; Resources exist to expose files. Reusing Resources inherits URI addressability, `resources/read`, resource subscriptions (the `resourceSubscriptions` filter on `subscriptions/listen`), and the existing client tooling for free. A new primitive would duplicate most of this and add ecosystem complexity. Using resources to describe files is also aligned with [Composability over specificity](https://modelcontextprotocol.io/community/design-principles#composability-over-specificity) and other MCP [design principles](https://modelcontextprotocol.io/community/design-principles).
```

Reviewer 2: @aditya-scio thinking about your backwards compatibility Q on this one re: resource subscriptions - I think this reference is another one that is more about the rationale

Existing implementations using other `skill://` URI structures will need to adjust to conform — see the Working Group's [related-work survey](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/related-work.md) for a catalog. Notably, FastMCP's widely-used [SkillsProvider](https://gofastmcp.com/servers/providers/skills) diverges on URI structure, discovery (per-skill `_manifest` vs. central index), and metadata mapping; coordinating that migration is a near-term Working Group priority. These are mechanical changes, not semantic ones.

## Security Implications

> Skill content is instructional text delivered to a model, which makes it a prompt-injection surface. The Working Group's position, recorded in [open-questions.md §10](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/open-questions.md#10-how-should-skills-handle-security-and-trust-boundaries), is:

Reviewer 9: open-questions.md section 10 records an open question plus a one-line proposed-mitigations note, not the normative list that follows the colon here, so the attribution reads as sourced when it isn't. The requirements are this extension's own:

```suggestion
Skill content is instructional text delivered to a model, which makes it a prompt-injection surface (background in [open-questions.md §10](https://github.com/modelcontextprotocol/experimental-ext-skills/blob/main/docs/open-questions.md#10-how-should-skills-handle-security-and-trust-boundaries)). This extension imposes the following requirements:
```

On top of that baseline, three mechanisms are defined. Two are discovery: enumeration via `skills/list`, which every server declaring this extension implements, and an optional pointer from server instructions. The third is retrieval: however a host arrives at a skill's URI, `skills/get` returns that skill's entry — its metadata and digests — including for skills no listing mentions.

#### Enumeration via `skills/list`

A server declaring the `io.modelcontextprotocol/skills` extension MUST implement the `skills/list` method, which returns the skills it serves. The result MAY be empty.

The request carries an optional pagination cursor:

```json
{
  "jsonrpc": "2.0",
  "id": 4,
  "method": "skills/list",
  "params": {}
}
```

The result carries the skill entries:

```json
{
  "jsonrpc": "2.0",
  "id": 4,
  "result": {
    "resultType": "complete",
    "skills": [
      {
        "uri": "skill://git-workflow/SKILL.md",
        "frontmatter": {
          "name": "git-workflow",
          "description": "Follow this team's Git conventions for branching and commits"
        },

>         "resources": [

Reviewer 11: I am not sure I understand the value for listing resources explicitly here -- IMO, it seems like the intended flow should be:
- skills/list provides the info necessary to list avaible skills
- skills/get provides the equivalent of activate_skill tool call -- the content in the skill (including the SKILL.md)

offering this info in skills/list feels like just asking for coordination/caching challenges.

Author: Skill contents (SKILL.md and other files) are always in resources and nowhere else. The intended flow is that a client calls `skills/list`, which provide all metadata and then when it comes time to activate a skill you do the `resources/read` for the contents. `skill/get` exists for when you need to either refresh the metadata from `skills/list` for a single skill (e.g. digest mismatch) or you need to know the metadata for an unlisted skill.

We could just have `skills/list` return a list of URIs and the clients would have to `skills/get` to get the resources and then `resources/read` for the content, but it's an extra hop on the activation path that we'd prefer to avoid.

**Context.** The host surfaces the `name` and `description` of each enabled registry entry in the model's context — the same list the model already sees for filesystem skills, now with MCP-served entries mixed in. The host's UI presents the same merged list for user inspection and per-skill enable/disable, with provenance shown so users can see which server a skill came from.

> **Loading.** The host exposes a single skill-loading tool to the model, keyed by skill name:

Reviewer 11: Doesn't this conflict with the guidance that it should accept a URI?

if I specify a URI in a different skill, how does the model make the gap from using that URI to access this skill by name? especially if the name might be modify to prevent duplicates by a host

Author: This is an implementation detail but usually the host will have a unique name for the skill. Claude Code does this for example and it is on them to make sure it is unique:

> When a name is collision-qualified ([Names](#names)), the qualified form is what appears in the model's context and what the model passes as `name`.

but I agree this is unclear since "skill name" here would more naturally mean "the name provided by the skill" and not the mangled skill name, so it does read as contradictory. I will change to recommend it is keyed by server ID and URI.

Reviewer 11: I guess I'm trying to reconcile this with skills/get and the idea that the skills may not be fully enumerated. 

My understanding is that the idea was that a skill could list another skill (e.g. use <migration-skill-uri> to switch from x to y), but that doesn't seem fully feasible if the name surfaced to the agent/model is different per client.

Reviewer 15: This is a problem with all primitives, though - I can already say "Call the `get_issue` tool after identifying relevant issues with the `list_issues` tool (for example), but in practice both of those tools will be prefixed with the configured server identifier in most agents. It's an important concern, but I'm not certain it's one that this SEP can do much about.

Reviewer 15: The [updated sketch](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2640/commits/a3e147ca2710f68214247aecc729731ee1ae8d03) is less ambiguous, but I think this will still turn out to be a practical challenge with host-specific implementation decisions either way, and I expect to see varying decisions across the ecosystem regarding it (following it is asking host applications with existing skill support to change their parameter schema just to accommodate MCP; some will, others won't).

- An empty directory yields an empty `resources` array.
- Pagination mirrors `resources/list`: when the result includes `nextCursor`, the client passes it back as `cursor` to retrieve the next page.

A server that declares `directoryRead` MUST support the method for every directory within the skill namespaces it serves as individual files. The method itself is not skill-specific: a server MAY support it on any directory resource it serves, under any scheme.

## Implementation Guidelines

The following are recommendations for interoperable implementations. They are not part of the normative specification.

### Hosts: End-to-End Integration

This section sketches one way a host might wire MCP-served skills into an existing skills implementation. It is illustrative, not prescriptive — hosts are free to structure tools, naming, and routing however suits their architecture. The goal is that an MCP-served skill flows through the same loading and reading mechanics as a filesystem skill — while remaining origin-tagged, per [Security Implications](#security-implications).

> **Registry.** At startup and on connection change, the host assembles a single internal skill registry from every origin it supports: filesystem skill directories, and `skills/list` results from each connected MCP server that declares the `io.modelcontextprotocol/skills` extension. Each registry entry records the skill's `name` and `description` (from the entry's `frontmatter`) and its origin — for a filesystem skill, the local directory; for an MCP skill, the server identity and the `SKILL.md` resource URI. Because names collide within and across origins ([Names](#names)), the registry keys entries by origin and name together, qualifying colliding names for display and invocation rather than dropping either entry.

Reviewer 16: This all makes sense, Caching becomes really important here to minimize startup costs, and etags can also help. I do think this will also cause folks to start asking for versions on Skills themselves to assist with not having to download these files every time, but outside the scope of MCP

Reviewer 17: In think the etag adjacent work we are doing in transports should be broadly able to cover this. 

The problem with versioning is perhaps that it is at least mildly in conflict with server deploy/versioning considering a key benefit of skills over MCP is that they can correctly reference the server they describe and be always up to date.

Then again if the broader concept of versioning arrives, skills being part of that makes complete sense.
